src/main/drivers/compass/compass_hmc5883l.c

```diff
@@ -145,80 +153,121 @@ static void hmc5883_extiHandler(extiCallbackRec_t* cb)
+static void hmc5883lConfigureDataReadyInterruptHandling(magDev_t* mag)
 {
 #ifdef USE_MAG_DATA_READY_SIGNAL
-
-    if (!(hmc5883InterruptIO)) {
+    if (mag->magIntExtiTag == IO_TAG_NONE) {
         return;
     }
+
+    const IO_t magIntIO = IOGetByTag(mag->magIntExtiTag);
+
 #ifdef ENSURE_MAG_DATA_READY_IS_HIGH
-    uint8_t status = IORead(hmc5883InterruptIO);
+    uint8_t status = IORead(magIntIO);
     if (!status) {
         return;
     }
 #endif
 
-    EXTIHandlerInit(&hmc5883_extiCallbackRec, hmc5883_extiHandler);
-    EXTIConfig(hmc5883InterruptIO, &hmc5883_extiCallbackRec, NVIC_PRIO_MAG_INT_EXTI, EXTI_Trigger_Rising);
-    EXTIEnable(hmc5883InterruptIO, true);
+#if defined (STM32F7)
+    IOInit(magIntIO, OWNER_COMPASS_EXTI, 0);
+    EXTIHandlerInit(&mag->exti, hmc5883_extiHandler);
+    EXTIConfig(magIntIO, &gmag->exti, NVIC_PRIO_MPU_INT_EXTI, IO_CONFIG(GPIO_MODE_INPUT,0,GPIO_NOPULL));
+#else
+    IOInit(magIntIO, OWNER_COMPASS_EXTI, 0);
+    IOConfigGPIO(magIntIO, IOCFG_IN_FLOATING);
+    EXTIHandlerInit(&mag->exti, hmc5883_extiHandler);
+    EXTIConfig(magIntIO, &mag->exti, NVIC_PRIO_MAG_INT_EXTI, EXTI_Trigger_Rising);
+    EXTIEnable(magIntIO, true);
+#endif
+#else
+    UNUSED(mag);
 #endif
 }
 
-static bool hmc5883lRead(int16_t *magData)
-{
-    uint8_t buf[6];
 #ifdef USE_MAG_SPI_HMC5883
-    bool ack = hmc5883SpiReadCommand(MAG_DATA_REGISTER, 6, buf);
-#else
-    bool ack = i2cRead(MAG_I2C_INSTANCE, MAG_ADDRESS, MAG_DATA_REGISTER, 6, buf);
+/*
+ * XXX Fixme
+ * HMC5983 datasheet states SPI mode is
+ *     SCK is high when CS is high
+ *     Data is sampled at the rising edge of SCK
+ * However, it seems that SCK condition is ignored (works with SCK low when CS is high).
+ * Nevertheless, it is nice to conform to the datasheet specification when per device SPI mode
+ * is implemented.
+ */
+static void hmc5883SpiInit(busDevice_t *busdev)
+{   
+    static bool hardwareInitialised = false;
+
+    if (hardwareInitialised) {
+        return;
+    }
+
+    IOInit(busdev->busdev_u.spi.csnPin, OWNER_COMPASS_CS, 0);
+    IOConfigGPIO(busdev->busdev_u.spi.csnPin, IOCFG_OUT_PP);
+
+    IOHi(busdev->busdev_u.spi.csnPin); // Disable
+
+    spiSetDivisor(busdev->busdev_u.spi.instance, SPI_CLOCK_STANDARD);
+
+    hardwareInitialised = true;
+}
 #endif
+
+static bool hmc5883lRead(magDev_t *magdev, int16_t *magData)
+{
+    uint8_t buf[6];
+
+    busDevice_t *busdev = &magdev->busdev;
+
+    bool ack = busReadRegisterBuffer(busdev, HMC58X3_REG_DATA, buf, 6);
+
     if (!ack) {
         return false;
     }
     // During calibration, magGain is 1.0, so the read returns normal non-calibrated values.
     // After calibration is done, magGain is set to calculated gain values.
 
-    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magGain[X];
-    magData[Z] = (int16_t)(buf[2] << 8 | buf[3]) * magGain[Z];
-    magData[Y] = (int16_t)(buf[4] << 8 | buf[5]) * magGain[Y];
+    magData[X] = (int16_t)(buf[0] << 8 | buf[1]) * magdev->magGain[X] / 256;
```

Comment:
This may be problematic - gain may be larger than 1.0 (> 256), `int16_t` will then overflow on saturated magnetometer.
It is quite extreme case and reading is unusable anyway (earth's magnetic field is too small to overflow), but it is better to handle it anyway.
IMO it is easier/enough to use saturation to `int16_t` range:

```
static int16_t parseMag(uint8* raw, int16_t gain) {
  int ret = (int16_t)(raw[0] << 8 | raw[1]) * gain / 256;
  return constrain(ret, INT16_MIN, INT16_MAX);
}

...
magData[X] = parseMag(buf + 0, magdev->magGain[X]);
magData[Z] = parseMag(buf + 2, magdev->magGain[Z]);
magData[Y] = parseMag(buf + 4, magdev->magGain[Y]);

Reply:
(and the problem is the same for other magnetometers)